### PR TITLE
`Adaptive learning`: Allow creating and updating prerequisite competencies

### PR DESCRIPTION
**Pull request by rstief**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
We want to refactor the prerequisite management with the goal of linking exercise units to prerequisites as pre-tests for a course. To support this, prerequisites were changed to be actual entities. Thus, we should also be able to create and update new prerequisites.

### Description
- Adds buttons to create and edit (update) prerequisites to the competency management
- Expands the prerequisite table in the competency management to include all new information.

### Steps for Testing
Required: 
- 1 instructor 

1. Go to course management -> competencies
2. Create new prerequisites
3. Edit existing prerequisites
4. Change a prerequisite title to equal one of the existing ones. After a short delay an error message should be displayed!
5. Always verify that the data is updated correctly!

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| competency-form.component.ts | 87.77% | ✅  |
| competency-management.component.ts | 93.49% | ✅  |
| competency.service.ts | 90.66% | ✅  |
| create-competency.component.ts | 97.61% | ✅  |
| competency-recommendation-detail.component.ts | 94.73% | ✅  |
| competency-search.component.ts | 100% | ✅ |
| import-competencies.component.ts | 90.9% | ✅  |
| import-course-competencies.component.ts | 98.59% | ✅  |
| import-prerequisites.component.ts | 91.66% | ✅  |
| create-prerequisite.component.ts | 100% | ✅ |
| edit-prerequisite.component.ts | 96.42% | ✅  |
| prerequisite-form.component.ts | 100% | ✅ |
| prerequisite.service.ts | 93.1% | ✅  |
| competency.model.ts | 88.33% | ✅  |
| standardized-competency.model.ts | 100% | ✅ |
| course.model.ts | 100% | ✅ |
| lecture-wizard-competencies.component.ts | 97.24% | ✅ |
| course-competencies.component.ts | 95.31% | ✅  |
| course-prerequisites-modal.component.ts | 94.73% | ✅  |

#### Server
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| PrerequisiteRepository.java | 100% | ✅  |
| CourseCompetencyRepository.java | 84% | ✅  |
| PrerequisiteService.java | 100% | ✅  |
| PrerequisiteRequestDTO.java | 100% | ✅  |
| PrerequisiteResource.java | 100% | ✅  |
| CompetencyResource.java | 94% | ✅  |

### Screenshots
Create/Edit View
image
Prerequisite List in the Competency Management View
image

## Summary by CodeRabbit

- **New Features**
  - Added methods to create and update prerequisites for courses.
  - Introduced new routes and components for managing prerequisites.
  - Added logic to display loading spinners and forms conditionally in the UI.
  - Implemented functionality to retrieve course competency titles.

- **Bug Fixes**
  - Improved form validation for competency titles and descriptions.

- **Refactor**
  - Renamed `CompetencyValidators` to `CourseCompetencyValidators`.
  - Updated default mastery threshold from 80 to 100.
  - Adjusted method signatures for better clarity and functionality.

- **Tests**
  - Added comprehensive unit tests for new prerequisites components and services.
  - Enhanced existing tests to include new validation scenarios and utilities.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/CompetencyResource.java [hunk 1]
```diff
@@ -537,6 +542,19 @@ public ResponseEntity<List<Competency>> generateCompetenciesFromCourseDescriptio
...
     }
 
+    /**
+     * GET courses/{courseId}/competencies/titles : Returns the titles of all course competencies. Used for a validator in the client
+     *
+     * @param courseId the id of the current course
+     * @return the titles of all course competencies
+     */
+    @GetMapping("courses/{courseId}/competencies/titles")
+    @EnforceAtLeastEditorInCourse
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CompetencyResource.java:552]: Can we allow this for all students in the course?

File: src/main/webapp/i18n/de/competency.json [hunk 2]
```diff
@@ -207,6 +214,21 @@
...
+            },
+            "create": {
+                "title": "Erstelle eine neue Voraussetzung",
+                "titlePlaceholder": "Gib der Voraussetzung einen Titel",
+                "softDueDateTooltip": "Empfehle einen Zeitpunkt zu dem die Studierenden diese Voraussetzung gemeistert haben sollten. Dies ist keine harte Frist.",
+                "suggestedTaxonomy": "Vorschlag",
+                "error": {
+                    "titleRequired": "Eine Voraussetzung braucht einen Titel",
+                    "titleMaxLength": "Der Titel einer Voraussetzung ist auf maximal {{max}} Symbole beschränkt",
+                    "descriptionMaxLength": "Die Beschreibung der Voraussetzung ist auf maximal {{max}} Symbole beschränkt",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/competency.json:226]:
```suggestion
                    "titleMaxLength": "Der Titel einer Voraussetzung ist auf maximal {{max}} Zeichen beschränkt",
                    "descriptionMaxLength": "Die Beschreibung der Voraussetzung ist auf maximal {{max}} Zeichen beschränkt",
```

File: src/main/webapp/i18n/en/competency.json [hunk 3]
```diff
@@ -190,10 +189,18 @@
                 }
             }
         },
+        "courseCompetency": {
+            "title": "Title",
+            "course": "Kurs",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/en/competency.json:194]:
```suggestion
            "course": "Course",
```

**Review by Jan-Thurner — CHANGES_REQUESTED** [review]
When editing a Prerequisite-Title to match an already existing prerequisite and I click cancel, the value is still being submitted. 
image

File: src/main/webapp/i18n/en/competency.json [hunk 4]
```diff
@@ -207,6 +214,21 @@
                 "success": "Imported {{ numPrerequisites }} prerequisites.",
                 "selectedTableHeader": "Selected Prerequisites",
                 "selectedTableEmpty": "No competencies selected to import as prerequisites"
+            },
+            "create": {
+                "title": "Create a new Prerequisite",
+                "titlePlaceholder": "Give the prerequisite a title",
+                "softDueDateTooltip": "Guide students by selecting when you expect them to have mastered this prerequisite. This is not a hard deadline.",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/en/competency.json:221]: Do we use `due date` or `deadline` for competencies in the translations?

File: src/main/webapp/i18n/de/competency.json (hunk 2 above)
Comment (JohannesStoehr) [reply src/main/webapp/i18n/de/competency.json:226]: Can we also mention the current number of characters?

File: src/test/java/de/tum/in/www1/artemis/competency/PrerequisiteUtilService.java [hunk 5]
```diff
@@ -59,4 +60,15 @@ public List<Prerequisite> createPrerequisites(Course course, int numberOfPrerequ
...
         return prerequisites;
     }
+
+    /**
+     * Creates a PrerequisiteRequestDTO from a prerequisite
+     *
+     * @param prerequisite the prerequisite to conver
+     * @return the created PrerequisiteRequestDTO
+     */
+    public PrerequisiteRequestDTO prerequisiteToRequestDTO(Prerequisite prerequisite) {
```
Comment (JohannesStoehr) [inline src/test/java/de/tum/in/www1/artemis/competency/PrerequisiteUtilService.java:70]: Don't we usually add a `of` method in the DTO for that?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by florian-glombik — CHANGES_REQUESTED** [review]
### Should be changed
* Using the cancel button should lead back to the Competencies overview, right now it 
noRedirectOnCancel

* Right now the Cancel button does always save

### Noticed but does not need to be changed
* For the title we do only have client validation, I think that we should never trust the client - but since it does not seem to be a hard requirement that the titles are different it does not need to be changed for me

**Review by Jan-Thurner — APPROVED** [review]
retested on TS2. Works as expected now.

**Review by JohannesWt — DISMISSED** [review]
Code Looks good

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by florian-glombik — APPROVED** [review]
Issue described in https://github.com/ls1intum/Artemis/pull/8765#pullrequestreview-2107422051 are fixed now, tested on ts2 - works as described

**Review by SimonEntholzer — DISMISSED** [review]
Tested on Ts3, and works as expected. Code lgtm too :+1:

**Review by sarpsahinalp — DISMISSED** [review]
Tested locally works as expected

File: src/test/javascript/spec/component/competencies/competency-form.component.spec.ts [hunk 6]
```diff
@@ -61,6 +61,17 @@ describe('CompetencyFormComponent', () => {
         const competencyService = TestBed.inject(CompetencyService);
         const getAllTitlesSpy = jest.spyOn(competencyService, 'getCourseCompetencyTitles').mockReturnValue(of(new HttpResponse({ body: ['test'], status: 200 })));
 
+        const competencyOfResponse: Competency = {};
```
Comment (Jan-Thurner) [inline src/test/javascript/spec/component/competencies/competency-form.component.spec.ts:64]: Is it possible to set the values `competencyOfResponse.id` and `competencyOfResponse.title` in the constructor already? In the same way as with the `response`?

```suggestion
        const competencyOfResponse: Competency = new competencyOfResponse({
                id: 1,
                title: 'test',
        });
```

File: src/test/javascript/spec/component/competencies/competency-form.component.spec.ts [hunk 7]
```diff
@@ -61,6 +61,17 @@ describe('CompetencyFormComponent', () => {
...
+        competencyOfResponse.id = 1;
```
Comment (Jan-Thurner) [inline src/test/javascript/spec/component/competencies/competency-form.component.spec.ts:65]:
```suggestion

```

File: src/test/javascript/spec/component/competencies/competency-form.component.spec.ts [hunk 8]
```diff
@@ -61,6 +61,17 @@ describe('CompetencyFormComponent', () => {
...
+        competencyOfResponse.title = 'test';
```
Comment (Jan-Thurner) [inline src/test/javascript/spec/component/competencies/competency-form.component.spec.ts:66]:
```suggestion

```

**Review by Jan-Thurner — APPROVED** [review]
Appart from the comments above, the new changes lgtm.

**Review by SimonEntholzer — APPROVED** [review]
Tested locally, and new changes lgtm.

File: src/test/javascript/spec/component/competencies/competency-form.component.spec.ts (hunk 6 above)
Comment (rstief) [reply src/test/javascript/spec/component/competencies/competency-form.component.spec.ts:64]: It is not completely possible, but we can do it like this:
```suggestion
        const competencyOfResponse: Competency = {
                id: 1,
                title: 'test',
        };
```

**Review by JohannesWt — APPROVED** [review]
Reapprove - Code

**Review by Jan-Thurner — DISMISSED** [review]
Code

**Review by Jan-Thurner — APPROVED** [review]
reapprove

**Review by JohannesStoehr — APPROVED** [review]
Reapprove

**Review by JohannesWt — APPROVED** [review]
Reapprove 👍

**Review by florian-glombik — APPROVED** [review]
Re-approve

**Review by SimonEntholzer — APPROVED** [review]
Re-approve

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**